azure-iot-device/azure/iot/device/common/version_compat.py

```diff
@@ -0,0 +1,37 @@
+# --------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+"""This module defines functions intended for providing compatibility between
+different versions of Python"""
+
+from six.moves import urllib
+
+
+def urlencode(query, quote_via=urllib.parse.quote_plus, safe=""):
```

Comment:
>urlencode [](start = 4, length = 9)

style nit: if this exists because we want a consistent way to encode for MQTT topics, would it be better to call this something like mqtt_encode and remove the two optional parameters?  This way we're clearly spelling out that URL-encoding rules and MQTT-topic-encoding rules are very different and the fact that we're using a URL-encode helper for encoding MQTT topic names gets hidden in this one place.

Reply:
The idea here is that this is literally the same as urllib.parse.urlencode, it just allows for us to use the "quote_via" parameter even in Python 2. Kind of similar to how we reimplement certain functions with the same name for async compatibility purposes in `async_compat.py`, the idea was here that we could reimplement for compatibility purposes across versions.

The idea of changing this to be tied to MQTT semantically is an interesting one though.

---
In reply to: [399663014](https://github.com/Azure/azure-iot-sdk-python/pull/514#discussion_r399663014) [](ancestors = 399663014)